x/auth/client/cli/decode.go

```diff
@@ -1,98 +1,56 @@
 package cli
 
 import (
-	"bytes"
 	"encoding/base64"
 	"encoding/hex"
-	"encoding/json"
-	"fmt"
 
 	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
 	"github.com/tendermint/go-amino"
 
 	"github.com/cosmos/cosmos-sdk/client"
 	"github.com/cosmos/cosmos-sdk/client/context"
-	"github.com/cosmos/cosmos-sdk/codec"
-	"github.com/cosmos/cosmos-sdk/version"
-	"github.com/cosmos/cosmos-sdk/x/auth/types"
 	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
 )
 
+const flagHex = "hex"
+
 // GetDecodeCommand returns the decode command to take Amino-serialized bytes
 // and turn it into a JSONified transaction.
 func GetDecodeCommand(codec *amino.Codec) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "decode [amino-byte-string]",
-		Short: "Decode an amino-encoded transaction string",
+		Short: "Decode an amino-encoded transaction string.",
 		Args:  cobra.ExactArgs(1),
-		RunE: func(cmd *cobra.Command, args []string) (err error) {
-			cliCtx := context.NewCLIContext().WithCodec(codec)
-
-			txBytes, err := base64.StdEncoding.DecodeString(args[0])
-			if err != nil {
-				return err
-			}
-
-			var stdTx authtypes.StdTx
-			err = cliCtx.Codec.UnmarshalBinaryLengthPrefixed(txBytes, &stdTx)
-			if err != nil {
-				return err
-			}
-
-			return cliCtx.PrintOutput(stdTx)
-		},
+		RunE:  runDecodeTxString(codec),
 	}
 
+	cmd.Flags().BoolP(flagHex, "x", false, "Treat input as hexadecimal instead of base64")
 	return client.PostCommands(cmd)[0]
 }
 
-// GetDecodeTxCmd - returns the command to decode a tx from hex or base64
-func GetDecodeTxCmd(cdc *codec.Codec) *cobra.Command {
-	return &cobra.Command{
-		Use:   "decode-tx [tx]",
-		Short: "Decode a tx from hex or base64",
-		Long: fmt.Sprintf(`Decode a tx from hex, base64.
-			
-Example:
-	$ %s tx decode-tx TWFuIGlzIGRpc3Rpbmd1aXNoZWQsIG5vdCBvbmx5IGJ5IGhpcyByZWFzb24sIGJ1dCBieSB0aGlz
-			`, version.ClientName),
-		Args: cobra.ExactArgs(1),
-		RunE: func(cmd *cobra.Command, args []string) error {
-
-			txString := args[0]
-
-			// try hex, then base64
-			txBytes, err := hex.DecodeString(txString)
-			if err != nil {
-				var err2 error
-				txBytes, err2 = base64.StdEncoding.DecodeString(txString)
-				if err2 != nil {
-					return fmt.Errorf(`expected hex or base64. Got errors:
-				hex: %v,
-				base64: %v
-				`, err, err2)
-				}
-			}
-
-			var tx = types.StdTx{}
-
-			err = cdc.UnmarshalBinaryLengthPrefixed(txBytes, &tx)
-			if err != nil {
-				return err
-			}
-
-			bz, err := cdc.MarshalJSON(tx)
-			if err != nil {
-				return err
-			}
-
-			buf := bytes.NewBuffer([]byte{})
-			if err = json.Indent(buf, bz, "", "  "); err != nil {
-				return err
-			}
-
-			fmt.Println(buf.String())
-			return nil
-		},
+func runDecodeTxString(codec *amino.Codec) func(cmd *cobra.Command, args []string) (err error) {
+	return func(cmd *cobra.Command, args []string) (err error) {
+		cmd.SetOut(cmd.OutOrStdout())
+		cmd.SetErr(cmd.ErrOrStderr())
```

Comment:
Just curious why these two calls are made. Why are they necessary? Should all commands do this? If so, why?

Reply:
Yes, unfortunately this calls should be made by all commands. I can address this in another PR though

Reply:
Fascinating. And why is it required? Note, not a single command does this atm. So you're saying all CLI commands are broken??

Reply:
> Yes, unfortunately this calls should be made by all commands to ensure stderr and stdout are redirected correctly. I can address this in another PR though

Is it possible to add them at the top level instead of on every cmd?

Reply:
Yes, I think so Fede. More changes will follow in ensuing PRs :+1:

Reply:
This:
```
cmd.SetErr(cmd.ErrOrStderr())
```

is required to ensure stderr is redirected to /dev/stderr correctly instead of /dev/stdout.
This is due to upstream's bad design choices. In order to avoid breaking backward compatibility, upstream decided to leave this sort of set up statements with the caller

Reply:
This should ideally be done for the command automatically. Why not do it when its all hooked up?

Reply:
Yes, can definitely do